### ISSUE TITLE
Error when predicting due to missing levels?

### ISSUE THREAD

**Issue by andreashandel** [mlr-org/mlr#811]
Hi,
I'm trying to use 10-fold cross validation on a simple linear model ("regr.lm") with the resample() function. I'm getting the following error message:
[Resample] cross-validation iter: 1
[Resample] cross-validation iter: 2
Error in model.frame.default(Terms, newdata, na.action = na.action, xlev = object$xlevels) : 
  factor XXX has new levels NN

(where XXX and NN can change depending on setting different seeds)

I assume it's because during the CV split, some factor levels are not in the training data and are dropped during the train part and then can't be used when predicting the holdout set, i.e. [like described in this stackoverflow topic.](http://stackoverflow.com/questions/4285214/predict-lm-with-an-unknown-factor-level-in-test-data)

Question: Is there some way to have resample() automatically deal with the error during prediction, e.g. similar to the answer to the stackoverflow question referenced above? Or other suggestions on how to best handle this within the mlr framework? (I want to eventually do feature selection and model tuning using CV with a few different methods).
Thanks!

**Comment by larskotthoff** [comment]
When creating the task, you can specify that empty levels shouldn't be dropped through `fixup.data`, e.g.

    task = makeClassifTask(id = "foo", data, fixup.data = "no")

**Comment by andreashandel** [comment]
Thanks for the quick reply! Tried it, does unfortunately not work. 
I think the problem is at the level of lm() itself, it discards the empty factors (see discussion in link above) and mlr would need to 'fix' what lm() produces as in the stackoverflow example before using predict().

**Comment by larskotthoff** [comment]
Ah, I see. That would mean removing data from the predictions though? I'm not sure if we want to do that in a predictor, as it would probably cause problems later on when something relies on predictions being there for all rows. I guess we could have a nicer error message though?

@berndbischl What's your take on this?

**Comment by andreashandel** [comment]
Wouldn't need to remove observations/rows from the prediction set, just setting factors in the prediction set that are not in training to NA. Would make for a worse prediction (as it should) but otherwise I think it could be ok. Should probably not be the default, but maybe on option one could turn on? 

The problem right now is that if I did a big tuning/feature selection/whatnot optimization and in just one of my many train/predict runs the training part of the sample by chance includes a factor level that is not in the prediction set, the whole thing likely fails (haven't tested, but I assume). Would be nice if mlr could somehow 'deal with' this one failure but still produce results for everything else.

**Comment by larskotthoff** [comment]
I see. It seems that another option (and probably preferable) would be to have an option to resample to stratify the factor levels such that all levels are present in all splits (if possible) though?

**Comment by andreashandel** [comment]
That sounds like a good option for some types of data. For my (small) dataset, I have a hunch that it might not work to get all factors for all variables present in both train and test. (I tried something like that with the stratify.cols option in makeResampleDesc(), it didn't work (could have been user error). 

And I'm also not sure how one would implement that for say bootstrap sampling without messing up the 'sample at random' statistical feature of the basic bootstrap?

**Comment by larskotthoff** [comment]
All good points.

Implementing this should be easy, but I'm a bit wary of the implications of changing the data to make something work technically.

**Comment by berndbischl** [comment]
@andreashandel 

mlr offers actually (hopefully) quite a lot of options to handle these problems. 
Do we have any chance to get a reproducible example from you, so we can easier discuss what helps you best?

**Comment by berndbischl** [comment]
I will outline different options now here:

1) stratified resampling. You can stratify your resampling procedure. have a look at makeResampleDesc args "stratify" (to stratify on y in classification) and "stratify.cols" (to stratify on imbalnced input features).
This might completely eliminate your problem, and I would do this if I can.
(Apparently you did this I see and it did not help you as the situation was more complex in your data)

2) Have a look at option "fix.factors.prediction" in makeLearners. 
From what I read in the stackoverflow thread, this does not solve your problem, but simpler one, but you should nevertheless know this option in this context.

3) For what is discussed in the stackoverflow thread: 
(I guess this is your real problem)
Summary might be: For some feature x_i a factor level occurs in prediction that was not seen in training. 
We dont have an option to handle this. Having one would be good. But we need to concrete protocol what to do for this.
We can of course predict NAs here. But would this help you? Not really, as you can still not compute a performance measure. So is this option still helpful? (IMHO: Maybe, but lets discuss this)

4) 
> The problem right now is that if I did a big tuning/feature selection/whatnot optimization and in just one of my many train/predict runs the training part of the sample by chance includes a factor level that is not in the prediction set, the whole thing likely fails (haven't tested, but I assume). Would be nice if mlr could somehow 'deal with' this one failure but still produce results for everything else.

Well, but we can handle exactly that, if this is indeed your problem.
Summary: Your model breaks, in only a few tuning iterations, due to WHATEVER reason (many exist).

a) configure your learner to "continue" if errors occur. see ?configureMlr with option on.learner.error. 
Note that this sets a global option (I guess I would do that for tuning), but you can also set it individually in makeLearner with arg "config".
This will now "on learner error" produce a FailureModel. Predictions will be NA. Measured performance will also be NA. 

b) look at TuneControl, option "impute.val" as some optimizers wont like the produced NA. 
We already have a reasobale default here.

Does any of this help? Do we need to improve docs?

**Comment by andreashandel** [comment]
@berndbischl 
@larskotthoff 

Thanks a lot for the quick and detailed reply!

I just started using mlr a few days ago and had not yet read the part in the documentation explaining the ability to set error handling with configureMlr(). I think that addresses my main concern, namely that I would start a large job and somewhere in there, just by chance something fails, e.g. due to CV not including a level in the train but in the predict set.

For my specific problem, I addressed it my further preprocessing/merging factors before doing the analysis. Is sufficient for this problem, I was more worried about the 'big picture' but that seems to be well addressed with configureMlr(). Getting a set of NA for a 'failed' run is I think perfect. Will alert me that something went wrong and then I can decide how to handle it.

Since you ask about improving docs: Overall I think they are great, but for some things I'd like to know a bit more. For instance, in caret (which is what I have mostly used, but got too frustrated with some of the lack of documentation/transparency - which is what led me to mlr) there is a list showing the exact names of the tuning parameters for each available model. I couldn't find something like that for mlr. For instance I was able to learn by looking at the tutorial examples that for randomForest one can tune ntree, but can't find any information if one can/can't tune mtry. Same for many other algorithms, I don't know where to find what I can tune through mlr.

Another one (and I should probably open a new issue for this, but I just add it here) is I can't find enough information in the tutorial to get pre-processing to work. I think I follow the tutorial instructions, which say to use preProcess syntax and add ppc in front, with this line:

makePreprocWrapperCaret(mylearner, ppc.method = c("center", "scale") )

But it fails with:
Error in UseMethod("paramValueToString") : 
no applicable method for 'paramValueToString' applied to an object of class "NULL"

It's probably me using the wrong syntax, but I can't find anywhere an example that shows how to do it.

Thanks for your help!

**Comment by schiffner** [comment]
> For instance I was able to learn by looking at the tutorial examples that for randomForest one can tune ntree, but can't find any information if one can/can't tune mtry. Same for many other algorithms, I don't know where to find what I can tune through mlr.

Generally, when interfacing a learner we try to take all its parameters into account, so usually you can tune every parameter you like.
The probably easiest way to get an overview of the parameters of a learner is via `getParamSet`.
For the random forest it looks like this:

```
getParamSet("classif.randomForest")
                    Type  len   Def   Constr Req Tunable Trafo
ntree            integer    -   500 1 to Inf   -    TRUE     -
mtry             integer    -     - 1 to Inf   -    TRUE     -
replace          logical    -  TRUE        -   -    TRUE     -
classwt    numericvector <NA>     - 0 to Inf   -    TRUE     -
cutoff     numericvector <NA>     -   0 to 1   -    TRUE     -
sampsize         integer    -     - 1 to Inf   -    TRUE     -
nodesize         integer    -     1 1 to Inf   -    TRUE     -
maxnodes         integer    -     - 1 to Inf   -    TRUE     -
importance       logical    - FALSE        -   -    TRUE     -
localImp         logical    - FALSE        -   -    TRUE     -
norm.votes       logical    -  TRUE        -   -   FALSE     -
do.trace         logical    - FALSE        -   -   FALSE     -
keep.inbag       logical    - FALSE        -   -   FALSE     -
```

The Tunable column indicates which parameters can be tuned. Except for the last three parameters, for which tuning does not make sense, all parameters can be tuned.

I could probably mention this on the tutorial page about tuning.

> Another one (and I should probably open a new issue for this, but I just add it here) is I can't find enough information in the tutorial to get pre-processing to work. I think I follow the tutorial instructions, which say to use preProcess syntax and add ppc in front, with this line:

> makePreprocWrapperCaret(mylearner, ppc.method = c("center", "scale") )

> But it fails with:
> Error in UseMethod("paramValueToString") : 
> no applicable method for 'paramValueToString' applied to an object of class "NULL"

> It's probably me using the wrong syntax, but I can't find anywhere an example that shows how to do it.

`makePreprocWrapperCaret` does not have a `method` argument as `preProcess`. Instead all preprocessing options that would go to the `method` argument of `preProcess` are individual logical parameters of `makePreprocWrapperCaret`.
In short, to make your example work, you need to do the following:

```
mylearner = makeLearner("classif.randomForest", ntree = 10)
lrn = makePreprocWrapperCaret(mylearner, ppc.center = TRUE, ppc.scale = TRUE )
train(lrn, iris.task)
```

**Comment by berndbischl** [comment]
> The Tunable column indicates which parameters can be tuned. Except for the last three parameters, for which tuning does not make sense, all parameters can be tuned.

Actually you COULD technically tell the tuner to tune those with "tunable = F", 
the meta info simply tells you, it does not make sense ;-)

Also note that you can even tune params that are NOT in the ParamSet of the learner (because we forgot to add them). See ?configureMlr, option "on.par.without.desc".

**Comment by berndbischl** [comment]
For the PreprocWrapperCaret:

Can we add an arg check or improve the docs here?

**Comment by schiffner** [comment]
> For the PreprocWrapperCaret:

> Can we add an arg check or improve the docs here?

Docs:
* The naming convention and what I wrote above about the `method` arg is mentioned in the tutorial (https://mlr-org.github.io/mlr-tutorial/devel/html/preproc/index.html#preprocessing-with-makepreprocwrappercaret). If there are any suggestions how to make this clearer I'm all ears.
* It's not in the docs of `makePreprocWrapperCaret` (see #408 ).

Arg check:
* The weird error message that @andreashandel saw originates from the print method for the wrapped learner because of a problem with `par.set` and `par.vals` I explained in #408 .

I can fix this next week.
@berndbischl: What's the status with mlr's own preprocessing wrapper?

**Comment by henningsway** [comment]
> For what is discussed in the stackoverflow thread:
> (I guess this is your real problem)
> Summary might be: For some feature x_i a factor level occurs in prediction that was not seen in training.
> We dont have an option to handle this. Having one would be good. But we need to concrete protocol what to do for this.
> We can of course predict NAs here. But would this help you? Not really, as you can still not compute a performance measure. So is this option still helpful? (IMHO: Maybe, but lets discuss this)

I am currently running into the same problem as I have a dataset with many kategorical predictors (some of them having some rare but important levels also). Any news on that front? :)

**Comment by berndbischl** [comment]
not yet, i think we should integrate package vtreat for this. 
i am currently writing up a gsoc proposal that contains this. 

at the moment i dont see a very simple solution

**Comment by pat-s** [comment]
Are you aware of the solutions posted after Sep 16? [Link](https://stackoverflow.com/a/44316204/4185785)

Whats your opinion on implementing this as an option for cases in which you have factor levels with n < n(folds)? 

`stratify.cols` solves the problem unless there are factor levels with the described problem OR if you have a spatial resampling (planned to be implemented) which you do want to preserve and not being altered by `stratify`.

**Comment by mb706** [comment]
`mlr` already does this, using the `fix.factors.prediction` flag:
```R
> ll = makeLearner("regr.lm", fix.factors.prediction=TRUE)
> t = train(ll, bh.task)
> newdata = getTaskData(bh.task, target.extra=T)$data[1:2, ]
> newdata
     crim zn indus chas   nox    rm  age    dis rad tax ptratio     b lstat
1 0.00632 18  2.31    0 0.538 6.575 65.2 4.0900   1 296    15.3 396.9  4.98
2 0.02731  0  7.07    0 0.469 6.421 78.9 4.9671   2 242    17.8 396.9  9.14
> newdata$chas  # the one factorial feature
[1] 0 0
Levels: 0 1
> newdata$chas = factor(c("0", "2"), levels=c("0", "1", "2"))
> newdata$chas
[1] 0 2
Levels: 0 1 2
> predict(t, newdata=newdata)
Prediction: 2 observations
predict.type: response
threshold: 
time: 0.01
  response
1 30.00384
2       NA
```

**Comment by pat-s** [comment]
Hi @mb706, 

to my understanding this fixes problems in which you have missing factor levels in `test` which are only present in `train`. In my case (and the case of OP) missing factor levels appear in `train` and are only present in `test` due to resampling. According to @berndbischl answer from March 30

> We dont have an option to handle this. Having one would be good. But we need to concrete protocol what to do for this.

The fix from [here](https://stackoverflow.com/a/44316204/4185785) could be also implemented under the `fix.factor.prediction` flag. It essentially sets factor levels in train, which are only present in `test`, to `NA` in `train` so that `predict()` will work. 

Correct me if I oversee something here and let me know what you are thinking about this :)

**Comment by mb706** [comment]
AFAICS your code solves a problem that is mostly already solved in `mlr`. `mlr` doesn't need to inspect the model to find out what features were present, it looks at the `TaskDesc` of the training task directly, which makes things more elegant. The "problem" currently is that `subsetTask`, which gets called for resampling, does not drop empty levels (as it arguably shouldn't).
The most direct route to making this work with resampling / subsetting of tasks would be to have `train` call `droplevels()` when `fix.factors.prediction` (or maybe some new flag) is `TRUE`, and hence get a `TaskDesc` which accurately names the factors used to train the model.

A [different kind of solution](https://rawgit.com/mlr-org/mlr/ComposablePreprocOperators/todo-files/CPO%20Vignette.html) (ctrl-F "cpoFixFactors") is also on the way btw (see my [PR](https://github.com/mlr-org/mlr/pull/1827)).

**Comment by pat-s** [comment]
AFAICS fixed by #2100. If not, we need a reprex.
